### PR TITLE
lxd: Send operation progress when creating image or container.

### PR DESCRIPTION
**Pull request by joelhockey**
Added metadata fields 'stage', 'percent', 'speed' to track
long running tasks in some operations.

In CreateImage, added stages 'create_image_from_container_tar'
and 'create_image_from_container_compress'.

In CreateContainerFromImage, added stage
'create_container_from_image_unpack'.

Modified images.go compressFile to take io.Reader and io.Writer
rather than a path name to allow imgPostContInfo to
supply an ioprogress.ProgressWriter that can track progress
of image compression.

Refactored imgPostContInfo to not require a separate pass
to determine the fingerprint when publishing an image from
a container.  The sha256 calculation is either done during
tar, or during compression.

Added tracker field in storageShared where callers can set
a ProgressTracker which is used when unpacking an image to
create new container in unpackImage and Unpack

Signed-off-by: Joel Hockey <joelhockey@chromium.org>

### CODE REVIEW

**Comment by monstermunchkin** [comment]
jenkins: test this please

File: lxd/images.go [hunk 1]
```diff
@@ -221,29 +228,66 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
+	tarfile.Close()
 
-	info.Size, err = io.Copy(sha256, tarf)
-	tarf.Close()
+	if usingCompression {
+		tarfile, err = os.Open(tarfile.Name())
+		if err != nil {
+			return nil, err
+		}
+		defer tarfile.Close()
```
Comment (monstermunchkin) [inline lxd/images.go]: Add blank line after this

File: lxd/images.go [hunk 2]
```diff
@@ -221,29 +228,66 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
+		fi, err := tarfile.Stat()
+		if err != nil {
+			return nil, err
+		}
```
Comment (monstermunchkin) [inline lxd/images.go]: Add blank line after this

File: lxd/images.go [hunk 3]
```diff
@@ -221,29 +228,66 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
+					metadata["speed"] = strconv.FormatInt(speed, 10)
+					op.UpdateMetadata(metadata)
+				},
+			},
+		}
+		compressedPath = tarfile.Name() + ".compressed"
+		compressed, err := os.Create(compressedPath)
+		if err != nil {
+			return nil, err
+		}
```
Comment (monstermunchkin) [inline lxd/images.go]: Add blank line after this

File: lxd/images.go [hunk 4]
```diff
@@ -221,29 +228,66 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
+		defer compressed.Close()
+		defer os.Remove(compressed.Name())
```
Comment (monstermunchkin) [inline lxd/images.go]: Add blank line after this

File: lxd/patches.go [hunk 5]
```diff
@@ -3154,7 +3154,17 @@ func patchMoveBackups(name string, d *Daemon) error {
 				}
 
 				// Compress it
-				compressedPath, err := compressFile(backupPath, "xz")
+				infile, err := os.Open(backupPath)
+				if err != nil {
+					return err
+				}
+				defer infile.Close()
```
Comment (monstermunchkin) [inline lxd/patches.go]: Add blank line after this

File: lxd/patches.go [hunk 6]
```diff
@@ -3154,7 +3154,17 @@ func patchMoveBackups(name string, d *Daemon) error {
...
+				compressed, err := os.Create(backupPath + ".compressed")
+				if err != nil {
+					return err
+				}
+				defer compressed.Close()
```
Comment (monstermunchkin) [inline lxd/patches.go]: Add blank line after this

File: shared/archive_linux.go [hunk 7]
```diff
@@ -73,8 +75,28 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
 		}
 		args = append(args, "-C", path, "--numeric-owner", "--xattrs-include=*")
 		args = append(args, extractArgs...)
-		args = append(args, file)
+		args = append(args, "-")
```
Comment (monstermunchkin) [inline shared/archive_linux.go]: Add blank line after this

File: shared/archive_linux.go [hunk 8]
```diff
@@ -73,8 +75,28 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
...
+		f, err := os.Open(file)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
```
Comment (monstermunchkin) [inline shared/archive_linux.go]: Add blank line after this

File: shared/archive_linux.go [hunk 9]
```diff
@@ -73,8 +75,28 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
...
+		reader = f
+		// Attach the ProgressTracker if supplied.
+		if tracker != nil {
+			fsinfo, err := f.Stat()
+			if err != nil {
+				return err
+			}
```
Comment (monstermunchkin) [inline shared/archive_linux.go]: Add blank line after this

File: lxd/images.go [hunk 10]
```diff
@@ -221,29 +228,66 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
+				Length: fi.Size(),
+				Handler: func(percent, speed int64) {
+					metadata["stage"] = "create_image_from_container_compress"
+					metadata["percent"] = strconv.FormatInt(percent, 10)
+					metadata["speed"] = strconv.FormatInt(speed, 10)
+					op.UpdateMetadata(metadata)
+				},
+			},
+		}
+		compressedPath = tarfile.Name() + ".compressed"
```
Comment (monstermunchkin) [inline lxd/images.go]: Add blank line after this

File: lxd/images.go [hunk 11]
```diff
@@ -221,29 +228,66 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
+		compressed, err := os.Create(compressedPath)
+		if err != nil {
+			return nil, err
+		}
+		defer compressed.Close()
+		defer os.Remove(compressed.Name())
+		// Calculate sha256 as we compress
+		writer := io.MultiWriter(compressed, sha256)
```
Comment (monstermunchkin) [inline lxd/images.go]: Add blank line after this

File: shared/archive_linux.go [hunk 12]
```diff
@@ -73,8 +75,28 @@ func Unpack(file string, path string, blockBackend bool, runningInUserns bool) e
...
 		args = append(args, "-C", path, "--numeric-owner", "--xattrs-include=*")
 		args = append(args, extractArgs...)
-		args = append(args, file)
+		args = append(args, "-")
+		f, err := os.Open(file)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+		reader = f
```
Comment (monstermunchkin) [inline shared/archive_linux.go]: Add blank line after this

File: lxd/images.go (hunk 4 above)
Comment (joelhockey) [reply lxd/images.go]: Done

File: lxd/images.go (hunk 3 above)
Comment (joelhockey) [reply lxd/images.go]: Done

File: lxd/images.go (hunk 2 above)
Comment (joelhockey) [reply lxd/images.go]: Done

File: lxd/images.go [hunk 13]
```diff
@@ -204,14 +192,33 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
+		WriteCloser: tarfile,
+		Tracker: &ioprogress.ProgressTracker{
+			Handler: func(percent, speed int64) {
+				metadata["stage"] = "create_image_from_container_tar"
+				metadata["percent"] = strconv.FormatInt(percent, 10)
+				metadata["speed"] = strconv.FormatInt(speed, 10)
+				op.UpdateMetadata(metadata)
+			},
+		},
+	}
```
[the remark this answers is not in the document]
Comment (joelhockey) [reply lxd/images.go]: Done

File: lxd/backup.go [hunk 14]
```diff
@@ -353,7 +353,17 @@ func backupCreateTarball(s *state.State, path string, backup backup) error {
...
+		infile, err := os.Open(backupPath)
+		if err != nil {
+			return err
+		}
+		defer infile.Close()
+		compressed, err := os.Create(backupPath + ".compressed")
+		if err != nil {
+			return err
+		}
+		defer compressed.Close()
```
[the remark this answers is not in the document]
Comment (joelhockey) [reply lxd/backup.go]: Done

File: lxd/backup.go [hunk 15]
```diff
@@ -353,7 +353,17 @@ func backupCreateTarball(s *state.State, path string, backup backup) error {
 	}
 
 	if compress != "none" {
-		compressedPath, err := compressFile(backupPath, compress)
+		infile, err := os.Open(backupPath)
+		if err != nil {
+			return err
+		}
+		defer infile.Close()
```
[the remark this answers is not in the document]
Comment (joelhockey) [reply lxd/backup.go]: Done

**Comment by stgraber** [comment]
jenkins: test this please

File: lxd/images.go [hunk 16]
```diff
@@ -221,29 +230,72 @@ func imgPostContInfo(d *Daemon, r *http.Request, req api.ImagesPost, builddir st
...
 	} else {
+		writer = io.MultiWriter(tarfileProgressWriter, sha256)
 		compressedPath = tarfile.Name()
 	}
-	defer os.Remove(compressedPath)
 
-	sha256 := sha256.New()
-	tarf, err := os.Open(compressedPath)
-	if err != nil {
+	if err := c.Export(writer, req.Properties); err != nil {
```
Comment (stgraber) [inline lxd/images.go]: Assignment and comparison should be done separately.

Comment (stgraber) [reply lxd/images.go]: (I realize this is existing code being moved around but just caught it now :))

**Comment by joelhockey** [comment]
Stéphane, I've split this into 5 different commits and made them separate pull requests.  I haven't used github much before, so I don't know if what I've sent is helpful.  I don't think there is a way for me to send the 5 different pull requests with dependencies on the previous.

I'll close this pull request now.

**Comment by stgraber** [comment]
Hi Joel,

I was thinking keeping a single pull request but splitting things into multiple commits inside it which makes for easier review and then cherry-picking to stable.

But multiple pull requests will work too.

Stéphane

**Pull request closed without merging**